Add edit functionality to environment and clients (#4887)
chef/automate: components/docs-chef-io/content/automate/chef_infra_server.md

An [environment]({{< relref "environments" >}}) can be used to map an organization's real-life workflow to what can be configured and managed when using Chef Infra. This mapping is accomplished by setting attributes and pinning cookbooks at the environment level.

The Chef Infra Server UI lets you:

- Create an environment.

> - Search for a specific environment from a list of environments

Reviewer: ```suggestion
- Search for a specific environment from a list of environments.
```

Reviewer 2: done

- List out all the environments at one place.
- View the details of an environment.
- Edit an environment.
- Delete an environment.

#### Create an Environment

To create a new Environment, select **Create Environment**, as shown below:

{{< figure src="/images/automate/create-environment-button.png" alt="Create Environment Button">}}

> Selecting the **Create Environment** button opens a dialog box. The dialog box contains four different sections named _Details_, _Constraints_, _Default Attributes_ and _Override Attributes_. Enter the details in the possible sections and select **Create** to create a new environment.

Reviewer: ```suggestion
Selecting the **Create Environment** button opens a dialog box containing four different sections, _Details_, _Constraints_, _Default Attributes_, and _Override Attributes_. Enter the details in the possible sections and select **Create** to create a new environment.
```

Reviewer 2: done

{{< figure src="/images/automate/create-environment-popup.png" alt="Create Environment Dialog Box">}}

#### Search for an Environment

> Use the **Search environments** bar to find an existing environment from the list of environments.
> Entering the name of an environment in the search box returns environments matching your search criteria.

Reviewer: ```suggestion
Use the **Search environments** field to search for any existing environment.
Automate will return any environments that match your search criteria.
```

Reviewer: Do we have any details on how it matches? For example glob matching?

{{< figure src="/images/automate/create-environment-button.png" alt="Create Environment Button">}}

#### Details of an Environment

> To view the details of an environment, follow the steps mentioned below:
> 
> - Select a specific environment.

Reviewer: ```suggestion
Select a specific environment to view that environment's cookbook constraints and attributes.
```

Reviewer 2: OK

>   - **Cookbook Constraints:** The selected bar contains the list of all the cookbook constraints with _Name_, their _Operator_, and _Version_ respectively. Select **Edit** to change the _Cookbook Constraints_.

Reviewer: This is a bit vague. I assume it is listing each cookbook in an environment by name, followed by the cookbook version number and the version constraint. We might want to link to https://docs.chef.io/cookbook_versioning/#constraints

```suggestion
**Cookbook Constraints**

The Cookbook Constraints window lists each cookbook by name followed by it's [version constraint]({{< relref "cookbook_versioning#constraints">}}) and version number.
```

  {{< figure src="/images/automate/cookbook-constraints-in-environment.png" alt="Cookbook Constraints In Environment">}}

>   - **Attributes:** The attributes bar shows all the default and overridden attributes. Select **Expand All** or **Collapse All** to view or hide the attributes. Select **Edit** to change the _Default Attributes_ and _Override Attributes_.

Reviewer: ```suggestion
**Attributes:** 

The attributes window shows all the default and overridden [attributes]({{< relref "attributes" >}}). Select **Expand All** or **Collapse All** to view or hide the attributes. Select **Edit** to change the _Default Attributes_ and _Override Attributes_.
```

Reviewer 2: Used h5's, but does the same thing.

>   {{< note >}} You can edit _Cookbook Constraints_ and the _Attributes_ only for environments with an existing _client_ for that _organization_.
>   {{< /note >}}

Reviewer: ```suggestion
{{< note >}} 
You can edit _Cookbook Constraints_ and the _Attributes_ only for environments with an existing _client_ for that _organization_.
{{< /note >}}
```

>   {{< figure src="/images/automate/attributes-in-environment.png" alt="Attributes In Environment">}}

Reviewer: ```suggestion
{{< figure src="/images/automate/attributes-in-environment.png" alt="Attributes In Environment">}}
```

Reviewer 2: done

#### Delete an Environment

> The Chef Infra Server lets you delete environments one at a time. Select **Delete** from the more information (three dots) icon on the far right side of the environment listing that you want to delete, as shown below:

Reviewer: ```suggestion
The Chef Infra Server lets you delete environments one at a time. Select **Delete** from the more information (three dots) icon on the far right side of the environment that you want to delete, as shown below:
```

Reviewer 2: done

{{< figure src="/images/automate/delete-an-environment.png" alt="Delete an Environment by selecting the more information icon on the right side of an environment list item">}}

### Data Bags

[Data bags]({{< relref "data_bags" >}}) store global variables as JSON data. Data bags are indexed for searching and can be loaded by a cookbook or accessed during a search.

### Clients

> Chef Infra Clients provide secure API access to the Chef Infra Server. Chef Infra Server UI lets you:

Reviewer: ```suggestion
Chef Infra Clients provide secure API access to the Chef Infra Server. The Chef Infra Server UI lets you:
```

- Create a client.
- Search for a specific client from a list of clients.
- List out all the clients at one place.
- Reset a client key.
- Delete a client.

#### Create a Client

To create a new client, select **Create Client**.

{{< figure src="/images/automate/create-client-button.png" alt="Create Client Button">}}

> Selecting the **Create Client** button opens a dialog box. The dialog box contains a section to add _Client Name_ and also an option to select whether the client is _Validation Client_ or not. Select **Create** .

Reviewer: What's the purpose of a Validation Client? This should probably link to docs that explain this so people know if they should check this or not. I'm assuming it has to do with `validation_client_name` here: https://docs.chef.io/config_rb_client/#settings

```suggestion
The **Create Client** button opens a dialog box. Enter the _Client Name_ and select _Validation Client_ or not. Select **Create** .
```

Reviewer 2: ok

{{< figure src="/images/automate/create-client-popup.png" alt="Create Client Dialog Box">}}

Selecting **Create** opens a dialog box which contains the _Private Key_ of that particular client. Select **Download** to download the _Private Key_.

{{< figure src="/images/automate/client-private-key.png" alt="Create Private Key of a Client">}}

#### Search for a Client

> Use the **Search Clients** bar to find an client from the list of clients.

Reviewer: ```suggestion
Use the **Search Clients** bar to find a client from the list of clients.
```

Entering the name of a client in the search bar returns clients matching your search criteria.

{{< figure src="/images/automate/create-client-button.png" alt="Create Client Button">}}

#### Public Key of a Client

> To view the _Public Key_ of a specific client, select the client. You will see the _Public Key_ of that client in **Details**.

Reviewer: ```suggestion
Select the client to view the _Public Key_ of a specific client. The _Public Key_ of that client will be displayed in **Details**.
```

Reviewer 2: ok

{{< figure src="/images/automate/client-details-public-key.png" alt="Public Key of Clients">}}

The Chef Infra Server lets you reset the _Public Key_ using the _Reset Key_ option. Selecting the _Reset Key_ option opens an alert stating _The current key will no longer be accepted._

{{< figure src="/images/automate/reset-key-in-client-details.png" alt="Reset Key option of a Client" width="400" height="300">}}

Selecting the _Reset Key_ shown in the above image opens a new dialog box which contains the new _Private Key_ of that client. To download the new _Private Key_, select **Download**.

> {{< figure src="/images/automate/reset-public-key-of-a-client.png" alt="Reset Public Key of a Clients">}}

Reviewer: ```suggestion
{{< figure src="/images/automate/reset-public-key-of-a-client.png" alt="Reset Public Key of a Client" width="500">}}
```

Reviewer 2: Let's discuss standards for images in our meeting

#### Delete a Client

> The Chef Infra Server lets you delete the created client one at a time. Select the **Delete** option from the right corner of the client you want to delete. Refer to the image below:

Reviewer: ```suggestion
Chef Infra Server lets you delete the existing clients one at a time. To delete a client, select the ellipses icon and then **Delete**. Refer to the image below:
```

Reviewer 2: ok

An [environment]({{< relref "environments" >}}) can be used to map an organization's real-life workflow to what can be configured and managed when using Chef Infra. This mapping is accomplished by setting attributes and pinning cookbooks at the environment level.

The Chef Infra Server UI lets you:

- Create an environment.

- Search for a specific environment from a list of environments

- List out all the environments at one place.

- View the details of an environment.

- Edit the Environment.

- Delete an environment.

#### Create an Environment

To create a new Environment, select **Create Environment**. Refer to the image below:

{{< figure src="/images/automate/create-environment-button.png" alt="Create Environment Button">}}

> Selecting the **Create Environment** button opens a popup menu. The popup menu contains four different sections named _Details_, _Constraints_, _Default Attributes_ and _Override Attributes_. Enter the details in the possible sections and select **Create** to create a new environment.

Reviewer 2: Review the Microsoft Style Guide for how to [describing interactions with the UI](https://docs.microsoft.com/en-us/style-guide/procedures-instructions/describing-interactions-with-ui) and revise the sections where you discuss  the "popup menu" accordingly.

  Using the **Edit** functionality, you can edit the _Cookbook Constraints_ for `n` number of times.

  - **Attributes:** The attributes bar shows all the default and overridden attributes. Select the **Expand All** or **Collapse All** option to view the attributes. Refer to the image below:

  {{< figure src="/images/automate/attributes-in-environment.png" alt="Attributes In Environment">}}

  Using the **Edit** functionality, you can edit the _Default Attributes_ and the _Override Attributes_ for `n` number of times.

#### Delete an Environment

The Chef Infra Server lets you delete the created environments one at a time. Select the **Delete** option from the right corner of the environment you want to delete. Refer to the image below:

{{< figure src="/images/automate/delete-an-environment.png" alt="Delete an Environment">}}

### Data Bags

[Data bags]({{< relref "data_bags" >}}) store global variables as JSON data. Data bags are indexed for searching and can be loaded by a cookbook or accessed during a search.

### Clients

Chef Infra Clients provide secure API access to the Chef Infra Server. Chef Infra Server UI lets you:

> - Create a client.

Reviewer 2: Single-space these bullet-points

An [environment]({{< relref "environments" >}}) can be used to map an organization's real-life workflow to what can be configured and managed when using Chef Infra. This mapping is accomplished by setting attributes and pinning cookbooks at the environment level.

The Chef Infra Server UI let's you access the _Environment_ object in following ways:

- All the _Environments_ of an Organization can be listed at one place.

- You can create a new Environment using the button as shown below:

{{< figure src="/images/automate/create-environment-button.png" alt="Create Environment Button">}}

> - In the above image, the search bar stating __Search environments__ can be used to search a specific environment from the list.

Reviewer 2: Please use `**` to bold words. It is a convention that helps anyone reading the markdown understand your intent.

Chef Infra Clients provide secure API access to the Chef Infra Server.

> The Chef Infra Server UI let's you access the _Clients_ object in the following ways:

Reviewer 2: `let's` should be `lets`

- All the _Clients_ of an organization can be listed at one place.

> - You can create a client using the button as shown below:

Reviewer 2: Instead of "as shown below" describe the action that the user takes

{{< figure src="/images/automate/create-client-button.png" alt="Create Client Button">}}

> - In the above image, the search bar stating __Search Clients__ can be used to search a specific client from the list.

Reviewer 2: I understand what you're trying to do with this list, but you're mixing conceptual and procedural content.

Conceptual:
The _Chef Infra Server_ tab lets you:
- Create a client
- Search for one client from a list of clients
- Delete a client
- View the details of a client
- Reset a client public key.

Then give the procedural steps for these actions. The sequence is "tell them what you are going to tell them how to do and then tell them how to do it."
